### PR TITLE
#4482 Correct the `DataTypeName` for PostgreSQL

### LINKED ISSUE #4482: Correct the `DataTypeName` for each DataBase
The definition of `DataTypeName` in `RuleName.g4`(Antlr grammer file) of each DataBase is not correct so far, therefore this issue is created to correct them. The sub-task list is below,
- [x] Correct the `DataTypeName` for `MySQL`.
- [x] Correct the `DataTypeName` for `PostgreSQL`. 
- [x] Correct the `DataTypeName` for `SQLServer`. 
- [x] Correct the `DataTypeName` for `Oracle`. 
- [x] Correct the `DataTypeName` for `SQL92`. 

Q1: How to learn about the definition of `DataTypeName`?
A:
 - Official document of each DataBase, i.g [MySQL](https://dev.mysql.com/doc/refman/8.0/en/data-types.html).
- [Antlr SQL grammer](https://github.com/antlr/grammars-v4/tree/master/sql)
- Other related Website. (I am not sure. 😃 )

Q2: Which file of ShardingSphere does `DataTypeName` exist?
A: All files named with `RuleName.g4`

### PR DESCRIPTION
**Pull request by yuzel**
Fixes #4482.

Changes proposed in this pull request:
-  add PostgreSQL dataType

Note:
- doublePrecision is support `double precision` . [DATATYPE-FLOAT](http://www.postgres.cn/docs/11/datatype-numeric.html#DATATYPE-FLOAT)
- identifier is support `complex type` [ROWTYPES-DECLARING](http://www.postgres.cn/docs/11/rowtypes.html#ROWTYPES-DECLARING)

### CODE REVIEW

File: shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/PostgreSQLVisitor.java [hunk 1]
```diff
@@ -460,6 +461,14 @@ public final ASTNode visitCaseExpression(final CaseExpressionContext ctx) {
     
     @Override
     public final ASTNode visitDataTypeName(final DataTypeNameContext ctx) {
+        DoublePrecisionContext doublePrecisionContext = ctx.doublePrecision();
+        if (null != doublePrecisionContext) {
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/PostgreSQLVisitor.java]: Here is a better way,
```
if (null != doublePrecisionContext) {
    return visit(ctx.doublePrecision());
}

@Override
public final ASTNode visitDoublePrecision(final DoublePrecisionContext ctx) {
    return new KeywordValue(xxxx);
}
```

File: shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/antlr4/imports/postgresql/BaseRule.g4 [hunk 2]
```diff
@@ -303,7 +303,14 @@ dataType
...
 
 dataTypeName
-    : identifier
+    : INT | SMALLINT | INTEGER | BIGINT | DECIMAL | NUMERIC | REAL | FLOAT | doublePrecision | SMALLSERIAL | SERIAL | BIGSERIAL
+    | MONEY | VARCHAR | CHARACTER | CHAR | TEXT | NAME | BYTEA | TIMESTAMP | DATE | TIME | INTERVAL | BOOLEAN | ENUM | POINT
+    | LINE | LSEG | BOX | PATH | POLYGON | CIRCLE | CIDR | INET | MACADDR | BIT | VARBIT | TSVECTOR | TSQUERY | UUID | XML
+    | JSON | INTRANGE | NUMRANGE | TSRANGE | TSTZRANGE | DATERANGE | ARRAY | identifier
+    ;
+
+doublePrecision
```
Comment (tristaZero) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/antlr4/imports/postgresql/BaseRule.g4]: Why does `doublePrecision` not exist in `PostgreSQLKeyword.g4 `?

Comment (yuzel) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/antlr4/imports/postgresql/BaseRule.g4]: > Why does `doublePrecision` not exist in `PostgreSQLKeyword.g4 `?

DOUBLE and PRECISION exist in KeyWord.g4, so I added an expression to represent double precision. Now I find that it is better to add keywords directly, I will change to keyword and remove doublePrecision.
`DOUBLEPRECISION
    : D O U B L E [ ]+ P R E C I S I O N
    ;`

**Comment by yuzel** [comment]
> I wonder whether you want to continue to involve in the community to do anything interesting or go on with #4482?

Yes, I want to continue to participate and contribute to the community. # 4482 I will continue to finish.

**Comment by tristaZero** [comment]
> Yes, I want to continue to participate and contribute to the community. # 4482 I will continue to finish.

@yuzel 
Very welcome. But my suggestion is to follow your own pace since I do not want this issue to stress you too much. 😃 

Grow up with the community and demonstrate your capabilities!

**Comment by TeslaCN** [comment]
`MONEY` is not a keyword of PostgreSQL.
https://github.com/postgres/postgres/blob/d952373a987bad331c0e499463159dd142ced1ef/src/include/parser/kwlist.h

**Pull request merged**